Gff3 Tribble Codec

kachulis opened the pull request:
Adds a Gff3 tribble codec for reading gff3 files.  Based on specification defined at https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md.  This PR focuses on implementing the capability to read this format; it does not implement any ontological validation of feature types or allowed relationships between features of different types.  It is thus only a partial implementation of the specification.

kachulis commented on src/main/java/htsjdk/tribble/gff/Gff3Codec.java:
```diff
@@ -0,0 +1,432 @@
+            return null;
+        }
+
+        public Object decode(final String line) throws IOException {
```
While that's true currently, I want to leave this as an Object because there are other directives in the spec that I haven't implemented at the moment, which might want to return other object types if implemented in the future.  For example the spec allows for directives indicating particular ontologies, or particular genome builds.

kachulis commented on src/main/java/htsjdk/tribble/gff/Gff3Codec.java:
```diff
@@ -0,0 +1,432 @@
+            final Set<Gff3Feature> parents = new HashSet<>();
+            for (final String parentID : parentIDs) {
+                final Set<Gff3Feature> theseParents = activeFeaturesWithIDs.get(parentID);
+                if (parents == null) {
```
Yes!

kachulis commented on src/main/java/htsjdk/tribble/gff/Gff3Codec.java:
```diff
@@ -0,0 +1,432 @@
+     * @param directiveLine
+     * @throws IOException
+     */
+    private void parseDirective(final String directiveLine) throws IOException {
```
The spec allows for arbitrary application specific directives, which particular parsers need not support.  Additionally there are some specified directives (particularly related to ontology definition) which aren't yet included in this implementation.  I think throwing on an unsupported directive is therefore too restrictive, so feel like logging a warning, while not great, is probably the best option.

lbergelson replied:
I see.  I was assuming there was a finite list of valid directives.  This makes sense in that case.

kachulis commented on src/main/java/htsjdk/tribble/gff/Gff3Codec.java:
```diff
@@ -0,0 +1,432 @@
+                break;
+
+            default:
+                throw new IllegalArgumentException( "directive " + directive + " not handled by parser.");
```
This exception is there to catch if a constant is added to the `Gff3Directive`  enum which then is not handled by this switch.  So I think `IllegalArgumentException` is more appropriate, since this exception is intended to catch the future introduction of a bug, not an issue with corrupt input data.  Perhaps an `AssertionError` or `IllegalStateException` would be better, my internet searches where inconclusive on the best practice for this type of situation.   I will improve the error message to make more explicit what's going on here.

kachulis commented on src/main/java/htsjdk/tribble/gff/Gff3Feature.java:
```diff
@@ -0,0 +1,417 @@
+            hash = 31 * hash + strand.hashCode();
+            hash = 31 * hash + attributes.hashCode();
+            if (ID != null) {
+                hash = 31 * hash + ID.hashCode();
```
nope, good catch!

kachulis commented on src/main/java/htsjdk/tribble/gff/Gff3Feature.java:
```diff
@@ -0,0 +1,417 @@
+                idsInLineage.contains(c.getAttribute(DERIVES_FROM_ATTRIBUTE_KEY))).
+                collect(Collectors.toList());
+        decendants.addAll(childrenToAdd);
+        idsInLineage.addAll(childrenToAdd.stream().map(Gff3Feature::getID).collect(Collectors.toSet()));
```
The modification does need to happen, since the Derives_from attribute could be pointing to a feature at any higher generation.  So as we move down through the generation, we need to keep adding to the list of ids in the particular lineage.  However, I can make this clearer (and reduce the side effect) by making a new list updated to include the new ids and call with that, instead of modifying the input list.

kachulis commented:
@lbergelson back to you.  While I was addressing your comments, I realized that I was depending on parent features being before child features in the input files, which is not at all guaranteed.  This is fixed now (with added tests), but means the parent maps, which used to be immutable, are now mutable, with public modifying methods.

lbergelson commented on src/main/java/htsjdk/tribble/gff/Gff3FeatureImpl.java:
```diff
@@ -172,17 +145,44 @@ public Gff3BaseData getBaseData() {
     public void addCoFeature(final Gff3FeatureImpl coFeature) {
+        if (!parents.equals(coFeature.getParents())) {
+
+            throw new TribbleException("Co-featrues " + baseData.id + " do not have same parents");
```
```suggestion
            throw new TribbleException("Co-features " + baseData.id + " do not have same parents");
```

kachulis commented:
@lbergelson a few changes since you approved, these resulted from getting some full scale use cases working:

1) clear `activeParentIDs` in `Gff3Codec::prepareToFlushFeatures()`.  This is needed to avoid memory leak and massive garbage collection problems when using (for example) a full scale human annotation file.

2) clear all maps and queues in `Gff3Codec::close`.  This is needed to avoid parsed, but unreturned features from persisting through `close`, and later being returned incorrectly when the same codec is used to read from a potentially completely different stream.

3) (minor) adjusted directive regex to allow for variable number of spaces 

At this point, my use cases are working, and I am content with where this is.  I don't have the authorization to merge this, so feel free to go ahead and merge if you don't have any issues with the additional changes.
